source/common/network/filter_manager_impl.h

```diff
@@ -112,6 +118,34 @@ class FilterManagerImpl : protected Logger::Loggable<Logger::Id::connection> {
   void onRead();
   FilterStatus onWrite();
   bool startUpstreamSecureTransport();
+  void maybeClose();
+  void onConnectionClose(ConnectionCloseAction close_action);
+
+  void finalizeClose(ConnectionCloseAction close_action) {
+    state_.local_close_pending_ = false;
+    state_.remote_close_pending_ = false;
+    connection_.closeConnection(close_action);
+  }
+
+  bool pendingClose() { return state_.local_close_pending_ || state_.remote_close_pending_; }
+
+protected:
+  struct State {
+    // Number of pending write filters awaiting closure.
+    uint32_t write_filter_pending_close_count_{0};
+
+    // Number of pending read filters awaiting closure.
+    uint32_t read_filter_pending_close_count_{0};
```

Comment:
Is necessary to use two different count for this? May single `pending_close_count_` is enough?

Reply:
removed, the original idea is to distinguish the read and write but it will be more complicated.